File: src/ee/facebookads/internal/FacebookRewardVideoAd.cpp
//
//  FacebookRewardVideoAd.cpp
//  ee_x
//
//  Created by Zinge on 10/11/17.
//
//

#include <cassert>

#include "ee/core/Logger.hpp"
#include "ee/core/IMessageBridge.hpp"
#include "ee/core/Utils.hpp"
#include "ee/facebookads/FacebookAdsBridge.hpp"
#include "ee/facebookads/internal/FacebookRewardVideoAd.hpp"
#include "ee/ads/internal/MediationManager.hpp"

namespace ee {
namespace facebook {
using Self = RewardedVideo;

namespace {
auto k__createInternalVideo(const std::string& id) {
    return "FacebookAds_createInternalVideo_" + id;
}

auto k__destroyInternalVideo(const std::string& id) {
    return "FacebookAds_destroyInternalVideo_" + id;
}

auto k__hasRewardedVideo(const std::string& id) {
    return "FacebookAds_hasRewardedVideo_" + id;
}

auto k__loadRewardedVideo(const std::string& id) {
    return "FacebookAds_loadRewardedVideo_" + id;
}

auto k__showRewardedVideo(const std::string& id) {
    return "FacebookAds_showRewardedVideo_" + id;
}

auto k__onRewarded(const std::string& id) {
    return "FacebookAds_Video_onRewarded_" + id;
}

auto k__onFailedToLoad(const std::string& id) {
    return "FacebookAds_Video_onFailedToLoad_" + id;
}
auto k__onLoaded(const std::string& id) {
    return "FacebookAds_Video_onLoaded_" + id;
}
auto k__onOpened(const std::string& id) {
    return "FacebookAds_Video_onOpened_" + id;
}
auto k__onClosed(const std::string& id) {
    return "FacebookAds_Video_onClosed_" + id;
}
} // namespace

Self::RewardedVideo(IMessageBridge& bridge, Logger& logger,
                    const std::string& adId)
    : Super(logger)
    , bridge_(bridge)
    , logger_(logger) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    adId_ = adId;

    bridge_.registerHandler(
        [this](const std::string& message) {
            onReward();
            return "";
        },
        k__onRewarded(adId_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            onFailedToLoad(message);
            return "";
        },
        k__onFailedToLoad(adId_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            onLoaded();
            return "";
        },
        k__onLoaded(adId_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            onOpened();
            return "";
        },
        k__onOpened(adId_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            onClosed();
            return "";
        },
        k__onClosed(adId_));
}

Self::~RewardedVideo() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    bridge_.deregisterHandler(k__onRewarded(adId_));
    bridge_.deregisterHandler(k__onFailedToLoad(adId_));
    bridge_.deregisterHandler(k__onLoaded(adId_));
    bridge_.deregisterHandler(k__onOpened(adId_));
    bridge_.deregisterHandler(k__onClosed(adId_));
}

bool Self::isLoaded() const {
    auto response = bridge_.call(k__hasRewardedVideo(adId_));
    return core::toBool(response);
}

void Self::load() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    bridge_.call(k__loadRewardedVideo(adId_));
}

bool Self::show() {
    if (not isLoaded()) {
        return false;
    }

    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto response = bridge_.call(k__showRewardedVideo(adId_));
    if (not core::toBool(response)) {
        return false;
    }

    auto&& mediation = ads::MediationManager::getInstance();
    auto successful = mediation.startRewardedVideo([this](bool rewarded) { //
        this->setResult(rewarded);

        this->destroyInternalVideo();
        this->createInternalVideo();
    });
    assert(successful);
    return true;
}

void Self::createInternalVideo() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto response = bridge_.call(k__createInternalVideo(adId_));
}

void Self::destroyInternalVideo() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto response = bridge_.call(k__destroyInternalVideo(adId_));
}
#pragma mark - on
void Self::onLoaded() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
}

void Self::onFailedToLoad(const std::string& message) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
}

void Self::onOpened() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    rewarded_ = false;
}

void Self::onReward() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    rewarded_ = true;
}

void Self::onClosed() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto&& mediation = ads::MediationManager::getInstance();
    auto successful = mediation.finishRewardedVideo(rewarded_);
    assert(successful);
}
} // namespace facebook
} // namespace ee

File: proj.android_studio/ee-x-facebook-ads/src/main/java/com/ee/facebook/FacebookRewardVideoAd.java
package com.ee.facebook;

import android.content.Context;
import android.support.annotation.NonNull;

import com.ee.core.Logger;
import com.ee.core.MessageBridge;
import com.ee.core.MessageHandler;
import com.ee.core.internal.Utils;
import com.facebook.ads.Ad;
import com.facebook.ads.AdError;
import com.facebook.ads.RewardedVideoAd;
import com.facebook.ads.RewardedVideoAdListener;

/**
 * Created by Zinge on 10/11/17.
 */

class FacebookRewardVideoAd implements RewardedVideoAdListener {

    private String k__createInternalVideo() {
        return "FacebookAds_createInternalVideo_" + _placementId;
    }

    private String k__destroyInternalVideo() {
        return "FacebookAds_destroyInternalVideo_" + _placementId;
    }

    private String k__hasRewardedVideo() {
        return "FacebookAds_hasRewardedVideo_" + _placementId;
    }

    private String k__loadRewardedVideo() {
        return "FacebookAds_loadRewardedVideo_" + _placementId;
    }

    private String k__showRewardedVideo() {
        return "FacebookAds_showRewardedVideo_" + _placementId;
    }

    private String k__onRewarded() {
        return "FacebookAds_Video_onRewarded_" + _placementId;
    }

    private String k__onFailedToLoad() {
        return "FacebookAds_Video_onFailedToLoad_" + _placementId;
    }

    private String k__onLoaded() {
        return "FacebookAds_Video_onLoaded_" + _placementId;
    }

    private String k__onOpened() {
        return "FacebookAds_Video_onOpened_" + _placementId;
    }

    private String k__onClosed() {
        return "FacebookAds_Video_onClosed_" + _placementId;
    }

    private static final Logger _logger = new Logger(FacebookRewardVideoAd.class.getName());

    private Context _context;
    private RewardedVideoAd _rewardedVideoAd;
    private String _placementId;

    public FacebookRewardVideoAd(@NonNull Context context, @NonNull String placementId) {
        Utils.checkMainThread();
        _context = context;
        _placementId = placementId;

        createInternalVideo();

        registerHandlers();
    }

    void destroy() {
        Utils.checkMainThread();
        deregisterHandlers();

        _context = null;
        _placementId = null;
    }

    private void registerHandlers() {
        Utils.checkMainThread();

        MessageBridge bridge = MessageBridge.getInstance();

        bridge.registerHandler(new MessageHandler() {
            @NonNull
            @Override
            public String handle(@NonNull String message) {
                loadRewardedVideo();
                return "";
            }
        }, k__loadRewardedVideo());


        bridge.registerHandler(new MessageHandler() {
            @NonNull
            @Override
            public String handle(@NonNull String message) {
                return Utils.toString(hasRewardVideo());
            }
        }, k__hasRewardedVideo());

        bridge.registerHandler(new MessageHandler() {
            @NonNull
            @Override
            public String handle(@NonNull String message) {
                return Utils.toString(showRewardVideo());
            }
        }, k__showRewardedVideo());

        bridge.registerHandler(new MessageHandler() {
            @NonNull
            @Override
            public String handle(@NonNull String message) {
                createInternalVideo();
                return "";
            }
        }, k__createInternalVideo());

        bridge.registerHandler(new MessageHandler() {
            @NonNull
            @Override
            public String handle(@NonNull String message) {
                destroyInternalVideo();
                return "";
            }
        }, k__destroyInternalVideo());
    }

    private void deregisterHandlers() {
        Utils.checkMainThread();
        MessageBridge bridge = MessageBridge.getInstance();

        bridge.deregisterHandler(k__hasRewardedVideo());
        bridge.deregisterHandler(k__loadRewardedVideo());
        bridge.deregisterHandler(k__showRewardedVideo());
    }

    private void createInternalVideo()
    {
        _rewardedVideoAd = new RewardedVideoAd(_context, _placementId);
        _rewardedVideoAd.setAdListener(this);
    }

    private void destroyInternalVideo()
    {
        _rewardedVideoAd.setAdListener(null);
        _rewardedVideoAd = null;
    }

    private boolean hasRewardVideo() {
        return _rewardedVideoAd.isAdLoaded();
    }

    private void loadRewardedVideo() {
        _rewardedVideoAd.loadAd();
    }

    private boolean showRewardVideo() {
        return _rewardedVideoAd.show();
    }

    @Override
    public void onRewardedVideoCompleted() {
        MessageBridge bridge = MessageBridge.getInstance();
        bridge.callCpp(k__onRewarded());
    }

    @Override
    public void onRewardedVideoClosed() {
        MessageBridge bridge = MessageBridge.getInstance();
        bridge.callCpp(k__onClosed());
    }

    @Override
    public void onAdClicked(Ad ad) {
        MessageBridge bridge = MessageBridge.getInstance();
        bridge.callCpp(k__onOpened());
    }

    @Override
    public void onError(Ad ad, AdError adError) {
        MessageBridge bridge = MessageBridge.getInstance();
        bridge.callCpp(k__onFailedToLoad());
    }

    @Override
    public void onAdLoaded(Ad ad) {
        MessageBridge bridge = MessageBridge.getInstance();
        bridge.callCpp(k__onLoaded());
    }

    @Override
    public void onLoggingImpression(Ad ad) {

    }
}
